### ISSUE TITLE
Can't download content on Windows

### ISSUE THREAD

**Issue by ryuheiyokokawa** [datproject/dat-node#238]
I am reporting:

- [X] a bug or unexpected behavior
- [ ] general feedback
- [ ] feature request
- [ ] security issue 

# Bug Report

Please give us details about your installation to assist you. Run `dat -v` to see the version of Dat you are using.

* Operating system: Win10 and Win7
* Node Version: 8.15 and 10.15
* dat-node Version: 3.5.15
* dat cli Version: 13.13 (used for testing)

### Expected behavior
Loaded dat address should download. (using example/download.js)

### Actual behavior
Connects, finds peers, then doesn't download content.
All this said, we know that version 3.5.11 works on Windows.  This is some sort of regression.

### Debug Logs
```
2019-04-24T02:10:19.183Z dat Dat DEBUG mode engaged, enabling quiet mode
2019-04-24T02:10:19.192Z dat dat 13.13.0
2019-04-24T02:10:19.192Z dat node v10.4.0
2019-04-24T02:10:19.510Z dat clone()
2019-04-24T02:10:19.512Z dat { _: null,
  utp: true,
  debug: false,
  quiet: true,
  sparse: false,
  upload: true,
  'show-key': false,
  k: false,
  logspeed: 400,
  empty: false,
  key: '<private>',
  dir: null,
  showKey: false }
2019-04-24T02:10:19.515Z dat Checking directory for clone
2019-04-24T02:10:19.516Z dat No existing directory, creating it.
2019-04-24T02:10:19.525Z dat-node archive ready. version: 0
2019-04-24T02:10:20.599Z dat archive syncing
2019-04-24T02:10:21.052Z dat-network Downloaded data: 284
2019-04-24T02:10:21.551Z dat-network Downloaded data: 284
2019-04-24T02:10:22.051Z dat-network Downloaded data: 284
2019-04-24T02:10:22.551Z dat-network Downloaded data: 284
2019-04-24T02:10:23.052Z dat-network Downloaded data: 284
2019-04-24T02:10:23.501Z dat-node Stream close
```

**Comment by ryuheiyokokawa** [comment]
Just to add some interesting facts/data to this, I'm finding that **dats hosted on a Windows machine does not work**.  I tested hosting dat on both Win7 and Win10 and it does not work. 

In the same environment, we tested all possible variations to/from Win/Lin/Mac.

Interestingly, the beaker version of "dat-node" can download Windows hosted dats no issue, except from Windows itself (touche).

**Comment by RangerMauve** [comment]
Weird, I've had Dat working on Windows for a few years now. :S

How many network interfaces do you have? I found I had some issues with that in the past.

**Comment by ryuheiyokokawa** [comment]
@RangerMauve thanks for dropping in!  Would you by chance be able to check the dat cli version you're using?  I'm curious as I think this is likely a regression on later versions. We experienced no issues earlier, but can't seem to find one that works well.

We're testing on a physical Windows 10 machine with a single ethernet.  I'm pretty sure we have only one IP assigned to it through DHCP on our local network.

We also tested from a Win10 and a Win7 VM inside Linux and Mac hosts.  They only have one bridged adapter (straight to ethernet on the host).  We've also tested this from outside the network Windows machines and they also experienced the same issues.

We're currently considering porting our code that uses this dat-node to the one from [beaker's version](https://github.com/beakerbrowser/dat-node).  That library needs some modifications, but it does work well to/from Windows.

**Comment by joehand** [comment]
Lets see if we can narrow this down to the kind of issue it is. Based on the logs, it seems like the networking is not the issue, is that correct? (or maybe that was tested locally)

The `examples/download.js` does not use the same storage as CLI. Does the CLI work for downloading files? Or is both that and the example broken.

**Comment by karissa** [comment]
I also wonder if you would be able to test 3.5.11, then 3.5.12, and 3.5.13, etc on your system to see in which version the regression is seen. This would help us narrow it down

**Comment by RangerMauve** [comment]
@ryuheiyokokawa I've bee using different projects like dat-gateway and beaker's dat APIs, so it's kinda hard to say which version I have. If none of the other stuff works, would you be able to create a repo with step-by-step guides for reproducing the issue?

**Comment by ryuheiyokokawa** [comment]
@joehand I'll look into the storage methods and see what's going on.  In our code we're using mirror/ram and that could be part of it?  I'll compare between the dat-node vs dat cli implementations.

@karissa Funny enough, I got 3.5.9 through 3.5.11 to work.  We'll continue testing up from there and see.

@RangerMauve We're trying a few options right now and will come back with a working/non-working gist

**Comment by ryuheiyokokawa** [comment]
@joehand @RangerMauve We've made some headway.  In our code (and maybe in the cli?), we had relied on the `dat.on('connection')` as a way to trigger the download method `dat.archive.metadata.update(download)`.  We now trigger the download method right away after the initial Dat callback (just like in the dat-node example download.js).  We think that this is the issue for **downloading dats to Windows**.  

@karissa we did test through lower and higher versions and didn't find that it fixed the issue.  We looked at older versions of our code and found that we were not wrapping the download method within the `on('connection')` event, thus making this look like a regression (which it may not be!).

In any case, the "on connection" wrapped code and cli works from both Linux and Mac.  
We're not ready to share a gist yet, but thank you again for the help!  We'll keep you updated on what we find.  We're still figuring out the **hosting dats from Windows** part.

**Comment by ryuheiyokokawa** [comment]
**More on hosting dats from Windows**
In addition, we found that the dat-node's `{temp:true}` option lets us create a clonable/syncable dat instance in Windows.  Hosting from windows works only when you have that option enabled.
We're working to figure out a way to get the `.dat` within the folder now.

**Comment by karissa** [comment]
thanks @ryuheiyokokawa, this is super useful

can you be clear on what you mean by 'hosting from windows works only when you have that option enabled' ? Curious what happens exactly (do you get an error?) Thanks!

**Comment by ryuheiyokokawa** [comment]
@karissa with the `temp:true` aka RAM, it somehow does alright (successful download on the otherside).  If we have the code write the .dat folder to disk (temp:false), then the other side can't download it.

Further, I decided to try sharing that folder through the CLI and run with `DEBUG=dat*`:
![dat-manager-test](https://user-images.githubusercontent.com/4183870/57562418-48a29e80-7347-11e9-9279-1f4d8a0b5569.png)
The important line above is the ENOENT line with the C:\ twice in the line....
This leads me to believe that somehow path isn't being resolved correctly on the share side (on the CLI and perhaps in the dat-node library too?)

**Comment by nharlow89** [comment]
I have not fully solved this one but I believe the issue resides in the content/metadata storage mechanism, in this case the `dat-storage` module. This is the only real difference between `temp` and disk storage and may explain temp (ram) works and storage does not. For reference, we have managed to get the `@beaker/dat-node` package to replicate successfully across all platforms. That package does not pass a custom storage object to hyperdrive.

Just to recap, we are unable to clone a dat stored on disk when it is created on Windows. Not even on the same machine it was created. However, we are able to pull in the metadata from anywhere without the actual content. We believe this rules out a networking issue. 

Also, I was experiencing the same issue on version 3.5.9 through 3.5.11.

**Comment by karissa** [comment]
These lines could be the issue: https://github.com/datproject/dat-storage/blob/master/index.js#L10-L21

Instead, can you try it using `path.join` instead of adding `.dat/` ?

**Comment by karissa** [comment]
Oh, it could also be that `/` should be using `path.sep` instead? https://github.com/random-access-storage/random-access-file/blob/master/index.js#L18

**Comment by nharlow89** [comment]
Verified that this was the issue. Current fix is to lock `random-access-file` to `v2.1.0` (bug introduced in `v2.1.1`) until the following pull request comes through: https://github.com/random-access-storage/random-access-file/pull/20

**Comment by moritonal** [comment]
Hi all, I'm on Windows and I'm facing a problem download, just checking here that this bug is the same before raising a new issue?

```
$Env:DEBUG="dat*"; dat clone dat://96d2bc************************fa12e323619bfd
  dat Dat DEBUG mode engaged, enabling quiet mode +0ms
  dat dat 13.13.1 +19ms
  dat node v11.12.0 +1ms
  dat clone() +0ms
  dat error reading dat.json key Error: directory must contain a dat.json
    at getDatJsonKey (C:\Users\User\AppData\Roaming\npm\node_modules\dat\src\commands\clone.js:134:13)
    at Object.clone [as command] (C:\Users\User\AppData\Roaming\npm\node_modules\dat\src\commands\clone.js:57:18)
    at doCb (C:\Users\User\AppData\Roaming\npm\node_modules\dat\node_modules\subcommand\index.js:63:19)
    at processTicksAndRejections (internal/process/next_tick.js:74:9)
    at process.runNextTicks [as _tickCallback] (internal/process/next_tick.js:51:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:865:11)
    at internal/main/run_main_module.js:21:11 +2ms
  dat { _: null,
  dat   utp: true,
  dat   debug: false,
  dat   quiet: true,
  dat   sparse: false,
  dat   upload: true,
  dat   'show-key': false,
  dat   k: false,
  dat   logspeed: 400,
  dat   empty: false,
  dat   key: '<private>',
  dat   dir: null,
  dat   showKey: false } +9ms
  dat Checking directory for clone +16ms
  dat-node archive ready. version: 11 +0ms
  dat archive syncing +0ms
  dat-node Replication error: ENOENT: no such file or directory, open 'C:\Users\User\96d2bca62****************************jhgfd\test.txt' +0ms
  dat-node Stream close +4ms
  dat-node Replication error: Not opened +1s
  dat-node Stream close +2ms
  dat-node Replication error: Not opened +1s
  dat-node Stream close +1ms
  dat-node Replication error: Not opened +1s
  dat-node Stream close +1ms
```

My fix is to run `dat` over the Ubuntu Sub-system for Windows, although if I'm honest I'm facing a "Could not resolve link" issue there as well in some situations (might raise that as a different issue and look into it).
